src/libcore/convert.rs

```diff
@@ -62,6 +102,10 @@ pub trait AsMut<T: ?Sized> {
 
 /// A conversion that consumes `self`, which may or may not be expensive.
 ///
+/// **Note:** these traits are for trivial conversion. **They must not fail**. If
+/// they can fail, use a dedicated method which return an `Option<T>` or
+/// a `Result<T, E>`.
```

Comment:
It may also be worth noting directly on this trait that it's unlikely that it should be implemented directly (because of `From`) and libraries should instead implement the `From` trait.